### DISCUSSION TITLE
Need help with dovecot's settings when using an S3 bucket via rclone for maildata directory

### DISCUSSION THREAD

**Discussion by jayknyn** [docker-mailserver/docker-mailserver#3301] [Q&A]
Hi, I've been successfully using DMS for the past few years and now I'm trying to adjust my setup to use an S3 bucket via [rclone mount](https://rclone.org/commands/rclone_mount/) for the `maildata` directory to provide data reliability/redundancy and no limits on storage. However, I'm running into errors from dovecot regarding "duplicate transaction logs". 

I'm able to mount the S3 bucket at the `maildata` path, start up DMS with no errors, and sync email (via imapsync) to the rclone mounted mail directory. Then when I log in via an IMAP client (roundcube), I get the below errors regarding "duplicate transaction logs". However, there are no duplicate dovecot logs at the listed file paths. This has something to do either with my rclone mount config or my dovecot config.

I have read through the [dovecot docs](https://doc.dovecot.org/configuration_manual/nfs) and it seems this issue is related to using  NFS and I presume virtual filesystems like fuse (what rclone mount uses) and dovecot's feature of "locking" files to prevent multiple sources from accessing a file at the same time.

This is my custom `/config/dovecot.cf` (using Traefik):
```
haproxy_trusted_networks = 172.18.0.0/16
haproxy_timeout = 3 secs
service imap-login {
  inet_listener imaps {
    haproxy = yes
    ssl = yes
    port = 10993
  }
}
mail_max_userip_connections = 100
default_process_limit = 200
mmap_disable = yes
mail_fsync = always
lock_method = dotlock
```

These are the dovecot error logs:
```
mx-archive  | [   INF   ]  archive.email.com is up and running
mx-archive  | May  1 10:11:03 archive postfix/master[1200]: daemon started -- version 3.5.17, configuration /etc/postfix
mx-archive  | May  1 10:11:12 archive dovecot: imap-login: Login: user=<admin@archive.email.com>, method=PLAIN, rip=104.36.110.103, lip=172.18.0.10, mpid=1219, TLS, session=<9WAkZaL6qNtoJG5n>
mx-archive  | May  1 10:11:12 archive dovecot: imap(admin@archive.email.com)<1219><9WAkZaL6qNtoJG5n>: Error: Transaction log /var/mail/archive.email.com/admin/dovecot.list.index.log: duplicate transaction log sequence (2)
mx-archive  | May  1 10:11:12 archive dovecot: imap(admin@archive.email.com)<1223><Q6kwZaL6uttoJG5n>: Error: Mailbox INBOX: Transaction log /var/mail/archive.email.com/admin/dovecot.index.log: duplicate transaction log sequence (2)
mx-archive  | May  1 10:11:13 archive dovecot: imap-login: Login: user=<admin@archive.email.com>, method=PLAIN, rip=104.36.110.103, lip=172.18.0.10, mpid=1229, TLS, session=<RXcyZaL6vNtoJG5n>
mx-archive  | May  1 10:11:13 archive dovecot: imap(admin@archive.email.com)<1223><Q6kwZaL6uttoJG5n>: Logged out in=76 out=720 deleted=0 expunged=0 trashed=0 hdr_count=0 hdr_bytes=0 body_count=0 body_bytes=0
mx-archive  | May  1 10:11:13 archive dovecot: imap(admin@archive.email.com)<1229><RXcyZaL6vNtoJG5n>: Error: Mailbox Drafts: Transaction log /var/mail/archive.email.com/admin/.Drafts/dovecot.index.log: duplicate transaction log sequence (2)
mx-archive  | May  1 10:11:13 archive dovecot: imap(admin@archive.email.com)<1229><RXcyZaL6vNtoJG5n>: Error: Mailbox Sent: Transaction log /var/mail/archive.email.com/admin/.Sent/dovecot.index.log: duplicate transaction log sequence (2)
mx-archive  | May  1 10:11:13 archive dovecot: imap(admin@archive.email.com)<1229><RXcyZaL6vNtoJG5n>: Error: Mailbox Junk: Transaction log /var/mail/archive.email.com/admin/.Junk/dovecot.index.log: duplicate transaction log sequence (2)
mx-archive  | May  1 10:11:14 archive dovecot: imap(admin@archive.email.com)<1229><RXcyZaL6vNtoJG5n>: Error: Mailbox Trash: Transaction log /var/mail/archive.email.com/admin/.Trash/dovecot.index.log: duplicate transaction log sequence (2)
```

The user (admin@) maildata directory:
![Screen Shot 2023-05-01 at 11 24 57](https://user-images.githubusercontent.com/45739598/235477092-f6e4824f-6986-48a2-aed9-6e2b44e0d086.png)

The user's .Trash directory:
![Screen Shot 2023-05-01 at 11 25 10](https://user-images.githubusercontent.com/45739598/235477114-d51abb3a-0a5e-4d43-ba5d-154fede14afa.png)

Other details:
- DMS version: 11.3.1
- S3 provider: Cloudflare R2
- rclone version: 1.61.1

<details>
  <summary>my rclone mount config</summary>

  ```
rclone mount coscrypt-mx: /opt/mx/maildata \
--config="/opt/rclone/rclone.conf" --cache-dir="/opt/rclone/cache" \
--log-level=DEBUG --log-file="/opt/rclone/mount.log" \
--uid=5000 --gid=5000 --umask=022 \
--allow-other --allow-non-empty --use-server-modtime \
--attr-timeout=10s \
--dir-cache-time=24h \
--vfs-read-chunk-size=0 \
--vfs-cache-max-size=5G \
--vfs-write-back=10m \
--vfs-fast-fingerprint \
--vfs-cache-mode=full \
--vfs-read-ahead=128M --buffer-size=32Mi --transfers=16 \
--s3-chunk-size=100M --s3-upload-cutoff=500M --s3-no-check-bucket
  ```
</details>

**Comment by polarathene** [comment]
> I have read through the [dovecot docs](https://doc.dovecot.org/configuration_manual/nfs) and it seems this issue is related to using NFS and I presume virtual filesystems like fuse (what rclone mount uses) and dovecot's feature of "locking" files to prevent multiple sources from accessing a file at the same time.

This was my guess too just by the discussion title alone :sweat_smile: 

Searching the error, there are [similar experiences tied to NFS usage](https://www.mail-archive.com/search?l=dovecot@dovecot.org&q=subject:%22dovecot.index.log%5C%3A+duplicate+transaction+log+sequence+%5C%283%5C%29%22&o=newest&f=1), and I assume [this is you](https://forum.rclone.org/t/need-help-with-rclone-mount-config-for-dovecot-email-directory-regarding-duplicate-dovecot-transaction-logs/37925). Then there is [this bug report](https://marc.info/?l=dovecot&m=149025894810118&w=2), fairly old but states they encountered the same error with local filesystems only. I also noticed [GlusterFS to have the same issue and experience corruption](http://savannah.nongnu.org/bugs/?24551), but that is a report from 2008. [This discussion heavily implies it's remote filesystems](https://dovecot.dovecot.narkive.com/2cseBLF1/index-errors-with-multiple-imap-client-access-in-2-2#post2).

Beyond that. With v12 of DMS we did make a Dovecot change to not use chroot. Not sure if that would affect this, but might be worth comparing to the 11.3.1 release you've used?

The [Dovecot docs for NFS](https://doc.dovecot.org/configuration_manual/nfs/) does call out early not to use the quota service which DMS has enabled by default (_try with it disabled via [`ENABLE_QUOTAS=0`](https://docker-mailserver.github.io/docker-mailserver/latest/config/environment/#enable_quotas)_). As you're apparently not using a clustered setup with Dovecots Director, the [bottom part of those docs](https://doc.dovecot.org/configuration_manual/nfs/#clustering-without-director) may also have relevant information worth investigating? (_[we use the `maildir` storage type](https://github.com/docker-mailserver/docker-mailserver/blob/bba72daedf4ecf03a80eea571160371191237863/target/dovecot/10-mail.conf#L30), so should be less prone to corruption concerns apparently_)

**Reply by polarathene** [reply]
The [Dovecot source code for this error](https://github.com/dovecot/core/blob/d378a1d5d3e8c6589664ce8d6ee6ee8063cbb6cd/src/lib-index/mail-transaction-log-file.c#L496-L502) does imply some form of corruption with the index however?

That advice at the bottom of their docs for managing the index may be what you need to avoid the problem? (_or could be related to NFS caching too_ :man_shrugging: )
